Added Environment

isc-etamarch opened the pull request:
Ask for environment in configuration; Add changing environment in settings; Show environment in UI

fixes (#124 )

isc-etamarch committed:
Added Environment

Ask for environment in configuration; Add changing environment in settings; Show environment in UI

isc-etamarch committed:
Fix failing unittest

isc-etamarch committed:
Merge branch 'main' into fix-124

isc-pbarton commented on cls/SourceControl/Git/Settings.cls:
```diff
@@ -110,12 +113,16 @@ Method %Save() As %Status
         set @storage@("settings", "user", $username, "basicMode") = ..basicMode
     } 
 
+    set ^%SYS("SystemMode") = ..environmentName
```
Nitpick: might prefer calling $SYSTEM.Version.SystemMode(environment name) instead of setting the global directly.

isc-pbarton commented on cls/SourceControl/Git/Utils.cls:
```diff
@@ -163,6 +163,16 @@ ClassMethod WarnInstanceWideUncommitted() As %Boolean
 
+ClassMethod EnvironmentName() As %String
+{
+    quit $get(@..#Storage@("settings","environmentName"),1)
```
Let's default to the empty string instead of 1 for the environment name

isc-etamarch committed:
Fix raised issues

Set environment lobal through method, default environment name to empty

isc-tleavitt commented on cls/SourceControl/Git/Utils.cls:
```diff
@@ -163,6 +163,16 @@ ClassMethod WarnInstanceWideUncommitted() As %Boolean
 
+ClassMethod EnvironmentName() As %String
+{
+    quit $get(@..#Storage@("settings","environmentName"),"")
```
IMO this shouldn't be in a separate global - let's just have this call $System.Version.SystemMode

isc-tleavitt commented on cls/SourceControl/Git/Utils.cls:
```diff
@@ -163,6 +163,16 @@ ClassMethod WarnInstanceWideUncommitted() As %Boolean
+
+ClassMethod IsLIVE() As %Boolean
+{
+    quit $get(@..#Storage@("settings","environmentName"),1)="LIVE"
```
Should call ..EnvironmentName() to reduce code duplication

isc-tleavitt commented on cls/SourceControl/Git/Settings.cls:
```diff
@@ -171,6 +178,13 @@ ClassMethod Configure() As %Boolean [ CodeMode = objectgenerator ]
+        } elseif ((propertyDef) && (propertyDef.Name = "environmentName")) {
+            do %code.WriteLine(" set list(1) = ""DEVELOPMENT""")
+            do %code.WriteLine(" set list(2) = ""TEST""")
+            do %code.WriteLine(" set list(3) = ""LIVE""")
```
Per https://docs.intersystems.com/irislatest/csp/documatic/%25CSP.Documatic.cls?LIBRARY=%25SYS&CLASSNAME=%25SYSTEM.Version#SystemMode "FAILOVER" is also an option - should add that.

isc-tleavitt commented on cls/SourceControl/Git/Settings.cls:
```diff
@@ -171,6 +178,13 @@ ClassMethod Configure() As %Boolean [ CodeMode = objectgenerator ]
+            do %code.WriteLine(" set list(2) = ""TEST""")
+            do %code.WriteLine(" set list(3) = ""LIVE""")
+            do %code.WriteLine(" set list(4) = """"")
+            do %code.WriteLine(" set value = """"")
```
Default value should be the current system mode

isc-tleavitt commented on cls/SourceControl/Git/Settings.cls:
```diff
@@ -110,12 +113,16 @@ Method %Save() As %Status
         set @storage@("settings", "user", $username, "basicMode") = ..basicMode
     } 
 
+    do $SYSTEM.Version.SystemMode(..environmentName)
```
Please specifically test that someone with just %Developer can set this

isc-etamarch replied:
$SYSTEM.Version.SystemMode does not work for someone with %Developer

isc-etamarch replied:
Should I revert to setting the global directly (through a private helper method)?

isc-tleavitt replied:
Setting the global directly isn't going to work either - I'd recommend just wrapping the whole thing in a Try/Catch and ignoring any errors.

isc-tleavitt replied:
@isc-etamarch Yes, it'll just be a no-op.

isc-tleavitt reviewed (changes requested):
My review is up - a few changes recommended

isc-etamarch committed:
Use $SYSTEM.Version.SystemMode

Use $SYSTEM.Version.SystemMode to deal with environment name
Add 'Failover' as an option

isc-etamarch committed:
Merge branch 'main' into fix-124

isc-etamarch committed:
Fix typo

Replace 'property' with 'EnvironmentName'

isc-tleavitt commented on cls/SourceControl/Git/Settings.cls:
```diff
@@ -182,8 +182,9 @@ ClassMethod Configure() As %Boolean [ CodeMode = objectgenerator ]
-            do %code.WriteLine(" set value = """"")
+            do %code.WriteLine(" set list(4) = ""FAILOVER""")
+            do %code.WriteLine(" set list(5) = """"")
+            do %code.WriteLine(" set value = inst.property")
```
This is wrong - should be `do %code.WriteLine(" set value = inst."_property)` - although actually, shouldn't be needed at all because we do that on line 162.

isc-tleavitt replied:
@isc-etamarch we overlapped on this - you were fixing as I commented, but my point still stands that we shouldn't need to set value here at all.

isc-etamarch committed:
Merge branch 'main' into fix-124

isc-etamarch committed:
Minor fixes

Wrap the environment setting in a try catch to avoid error if user does not have enough priveledges
